finality-aleph/src/data_io/data_store.rs

```diff
@@ -0,0 +1,655 @@
+
+        use crate::data_io::proposal::{PendingProposalStatus::*, ProposalStatus::*};
+        match new_status {
+            Pending(PendingTopBlock) => {
+                // We register only a finality trigger, since a block import trigger has been already registered
+                // when the proposal was added.
+                self.register_next_finality_trigger(proposal);
+                false
+            }
+            Pending(TopBlockImportedButIncorrectBranch) => {
+                // We do nothing, this is a result of malicious behaviour. This proposal can still get through,
+                // but we wait till maintenance.
+                false
+            }
+            Pending(TopBlockImportedButNotFinalizedAncestor) => {
+                // This is possible in honest scenarios -- this node must be behind in finalization.
+                self.register_next_finality_trigger(proposal);
+                false
+            }
+            Finalize | Ignore => {
+                self.on_proposal_available(proposal);
+                true
+            }
+        }
+    }
+
+    // Outputs the current status of the proposal based on the `old_status` (for optimization).
+    fn check_proposal_availability(
+        &mut self,
+        proposal: &AlephProposal<B>,
+        old_status: Option<&ProposalStatus>,
+    ) -> ProposalStatus {
+        if let Some(status) = self.available_proposals_cache.get(proposal) {
+            return status.clone();
+        }
+        let status = get_proposal_status(&mut self.chain_info_provider, proposal, old_status);
+        if status == ProposalStatus::Finalize || status == ProposalStatus::Ignore {
+            // We can cache only if the proposal is available. If it is pending, its
+            // status might change and we should not recover it from the cache.
+            self.available_proposals_cache
+                .put(proposal.clone(), status.clone());
+        }
+        status
+    }
+
+    // For a proposal that might be new or not, check if it is available. If it is a new proposal
+    // and it is not available:
+    // 1) create an entry in pending_proposals
+    // 2) link this proposal to the input message by modifying the new added entry in pending proposals
+    //    and the provided message_info.
+    // 3) register an appropriate event trigger (for block import and/or finality).
+    // If the proposal is available, message_info is not modified.
+    fn add_message_proposal_dependency(
+        &mut self,
+        proposal: &AlephProposal<B>,
+        message_info: &mut PendingMessageInfo<B, Message>,
+        id: MessageId,
+    ) {
+        if !self.pending_proposals.contains_key(proposal) {
+            use crate::data_io::proposal::{PendingProposalStatus::*, ProposalStatus::*};
+            let status = self.check_proposal_availability(proposal, None);
+            match &status {
+                Pending(pending_status) => {
+                    match pending_status {
```

Comment:
Nit: Consider flattening these `match`es, since nothing happens directly on `Pending(_)`? I think  this would be slightly more readable.

Reply:
Done